cli/commands/asset/add.go

```diff
@@ -0,0 +1,175 @@
+)
+
+var rename string
+
+// AddCommand adds c ommand that allows user to add assets from Bonsai.
+func AddCommand(cli *cli.SensuCli) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "add [NAME]",
+		Short: "adds an asset definition fetched from Bonsai",
+		RunE:  addCommandExecute(cli),
+	}
+
+	cmd.Flags().StringVarP(&rename, "rename", "r", "", "rename the asset to the provided string after fetching it from Bonsai")
+
+	return cmd
+}
+
+func addCommandExecute(cli *cli.SensuCli) func(cmd *cobra.Command, args []string) error {
+	return func(cmd *cobra.Command, args []string) error {
+		// If no name is present print out usage
+		if len(args) != 1 {
+			_ = cmd.Help()
+			return errors.New("invalid argument(s) received")
+		}
+
+		name := args[0]
+
+		bAsset, err := corev2.NewBonsaiBaseAsset(name)
+		if err != nil {
+			return err
+		}
+
+		var version *goversion.Version
+		if bAsset.Version != "" {
+			version, err = goversion.NewVersion(bAsset.Version)
+			if err != nil {
+				return err
+			}
+		}
+
+		bonsaiClient := bonsai.New(bonsai.BonsaiConfig{})
+		bonsaiAsset, err := bonsaiClient.FetchAsset(bAsset.Namespace, bAsset.Name)
+		if err != nil {
+			return err
+		}
+
+		if version == nil {
+			fmt.Println("no version specified, using latest:", bonsaiAsset.LatestVersion())
+			version = bonsaiAsset.LatestVersion()
+		} else if !bonsaiAsset.HasVersion(version) {
+			availableVersions := bonsaiAsset.ValidVersions()
+			sort.Sort(goversion.Collection(availableVersions))
+			availableVersionStrs := []string{}
+			for _, v := range availableVersions {
+				availableVersionStrs = append(availableVersionStrs, v.String())
+			}
+			return fmt.Errorf("version \"%s\" of asset \"%s/%s\" does not exist\navailable versions: %s",
+				version, bAsset.Namespace, bAsset.Name, strings.Join(availableVersionStrs, ", "))
+		}
+
+		fmt.Printf("fetching bonsai asset: %s/%s:%s\n", bAsset.Namespace, bAsset.Name, version)
+
+		asset, err := bonsaiClient.FetchAssetVersion(bAsset.Namespace, bAsset.Name, version.String())
+		if err != nil {
+			return err
+		}
+
+		resources, err := ParseResources(asset)
+		if err != nil {
+			return err
+		}
+		if err := create.ValidateResources(resources, cli.Config.Namespace()); err != nil {
+			return err
+		}
+		for i := range resources {
+			meta := resources[i].Value.GetObjectMeta()
+			if rename != "" {
+				meta.Name = rename
+			} else {
+				meta.Name = fmt.Sprintf("%s/%s", bAsset.Namespace, bAsset.Name)
+			}
+			resources[i].Value.SetObjectMeta(meta)
+		}
+		if err := create.PutResources(cli.Client, resources); err != nil {
+			return err
+		}
+
+		fmt.Printf("added asset: %s/%s:%s\n", bAsset.Namespace, bAsset.Name, version)
+		return nil
+	}
+}
+
+// ParseResources is a rather heroic function that will parse any number of valid
+// JSON resources. Since it attempts to be intelligent, it likely contains bugs.
+//
+// TODO(amdprophet): This is a modified copy of the ParseResources function
```

Comment:
Out of curiosity, why did we have to modify this function?

Reply:
`ParseResources` in the `create` command package takes in an `io.Reader` to read a file. I had to create a modified copy of the function to accept a string (of JSON). We also only need to support JSON format in this case.

Reply:
If indeed that's exactly the same function with a `string` parameter instead of an `io.Reader`, you can avoid duplication by putting the string into something that meets that interface, eg:

```go
// This gives you a bytes.Reader, which meets, among others, the io.Reader interface
bytes.NewReader([]byte("some string"))
```

Would that be enough to solve your issue?